### ISSUE TITLE
Axioms are restricted to Atoms

### ISSUE THREAD

**Issue by quicquid** [gapt/gapt#447]
In LKsk, initial sequents F :- F can contain arbitrary formulas F  - this is currently impossible in LK:

```
gapt> val f = And(Eq(FOLConst("a"), FOLConst("b")), Top())
f: at.logic.gapt.expr.FOLFormula = (a=b∧⊤)

gapt> Axiom(List(f),List(f))
java.lang.IllegalArgumentException: Cannot create axiom from sequent (a=b∧⊤) :- (a=b∧⊤).
  at at.logic.gapt.proofs.lkNew.Axiom$.apply(lkNew.scala:280)
  at at.logic.gapt.proofs.lkNew.Axiom$.apply(lkNew.scala:290)
  ... 30 elided
```

@gebner Is there are reason for the restriction to atoms? Currently it is also impossible to start a proof with arbitrary sequents, so there is no alternative.

**Comment by gebner** [comment]
1) Please don't use `Axiom`; it is a fairly fragile wrapper which should be avoided IMHO.  The actual inferences are called `LogicalAxiom`, `ReflexivityAxiom`, `TopAxiom`, `BottomAxiom`, and `TheoryAxiom`.

2) You can simulate a non-atomic logical axiom with `AtomicExpansion`:
```
scala> AtomicExpansion(f)
res1: at.logic.gapt.proofs.lkNew.LKProof =
[p5] (a=b∧⊤) :- (a=b∧⊤)    (AndLeftRule(p4, Ant(0), Ant(1)))
[p4] a=b, ⊤ :- (a=b∧⊤)    (AndRightRule(p3, Suc(0), p2, Suc(0)))
[p3] a=b :- a=b    (LogicalAxiom(a=b))
[p2] ⊤ :- ⊤    (WeakeningLeftRule(p1, ⊤))
[p1]  :- ⊤    (TopAxiom())
```

3) For `LogicalAxiom` I believe the restriction is more a matter of principle: inferences should be as basic as possible, and we can get away with only atomic axioms.  For `TheoryAxiom` we allow arbitrary clauses, here I believe the restriction was motivated by CERES, which can handle atomic initial sequents?  I like the restriction to atomic `LogicalAxioms` but I'm not particularly attached to the restriction in `TheoryAxiom`.

Do you have a particular use case for non-atomic initial sequents?

**Comment by quicquid** [comment]
Ah ok, I matched the comment for top to TheoryAxiom (i.e. i assumed it only allowed sequents of the form :- T, where T is a formula )

Off the top of my hat, I know two reasons:

* It blows up the proof size: each expansion introduces 2*n inferences, where n is the number of logical operators in the formula.

* In higher-order logic, the clause normal form is not closed under substitution. (E.g. substitute X <- (\forall x.x=x) in X :- X ). 

That's the reason why the CNF transformation is part of the Ral rules - we need to dynamically deconstruct such formulas, not only as a preprocessing step. This is also the reason, why the algorithm on page 13 on the CERESomega paper shows how to transform a proof starting with the axiom -A :- -A into a proof starting with A :- A etc.

In principle, we could have these general axioms in LKsk only, and use TheoryAxioms in LK for that, but i find it a bit confusing.

**Comment by gebner** [comment]
> In higher-order logic, the clause normal form is not closed under substitution. (E.g. substitute X <- (\forall x.x=x) in X :- X ).

Hmm, this is a real problem, but not because of the CNF.  It means that currently higher-order LK proofs are not closed under substitution...

> This is also the reason, why the algorithm on page 13 on the CERESomega paper shows how to transform a proof starting with the axiom -A :- -A into a proof starting with A :- A etc.

Ah thanks for explaining that, I was completely lost and wondered where the non-atomic axioms came from, but now I realize that this is an invariant of the algorithm.

I'd suggest we discuss this topic tomorrow at the gapt meeting.  I have run into similar problems when eliminating atom definitions as well.

**Comment by quicquid** [comment]
Yes - in principle you could replace D :- D by the atomic expansion of D after definition elimination. In case of theories, I'd say the axioms are usually non-atomic, even if they can often be written as sequents with atoms.

**Comment by gebner** [comment]
Solved by merging #449, logical axioms can now contain arbitrary formulas, not just atoms.
